sdk/examples/devmode_rust/src/engine.rs

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::sync::mpsc::{Receiver, RecvTimeoutError};
+use std::thread::sleep;
+use std::time;
+
+use rand;
+use rand::Rng;
+
+use sawtooth_sdk::consensus::{engine::*, service::Service};
+
+const DEFAULT_WAIT_TIME: u64 = 0;
+
+pub struct DevmodeService {
+    service: Box<Service>,
+}
+
+impl DevmodeService {
+    pub fn new(service: Box<Service>) -> Self {
+        DevmodeService { service }
+    }
+
+    fn wait_for_chain_head(&mut self) -> Block {
+        let mut query_result = self.service.get_chain_head();
+
+        while let Err(Error::NoChainHead) = query_result {
+            warn!("Waiting for chain head");
+            sleep(time::Duration::from_millis(200));
+            query_result = self.service.get_chain_head();
+        }
+
+        query_result.expect("Failed to get chain head")
+    }
+
+    fn get_chain_head(&mut self) -> Block {
+        debug!("Getting chain head");
+        self.service
+            .get_chain_head()
+            .expect("Failed to get chain head")
+    }
+
+    fn get_block(&mut self, block_id: BlockId) -> Block {
+        debug!("Getting block {:?}", block_id);
+        self.service
+            .get_blocks(vec![block_id.clone()])
+            .expect("Failed to get block")
+            .remove(&block_id)
+            .unwrap()
+    }
+
+    fn initialize_block(&mut self) {
+        debug!("Initializing block");
+        self.service
+            .initialize_block(None)
+            .expect("Failed to initialize");
+    }
+
+    fn finalize_block(&mut self) {
+        debug!("Finalizing block");
+        let mut query_result = self.service.finalize_block(Vec::from(&b"Devmode"[..]));
+
+        while let Err(Error::BlockNotReady) = query_result {
+            warn!("Block not ready");
+            sleep(time::Duration::from_secs(1));
+            query_result = self.service.finalize_block(Vec::from(&b"Devmode"[..]));
+        }
+
+        query_result.expect("Failed to finalize block");
+    }
+
+    fn check_block(&mut self, block_id: BlockId) {
+        debug!("Checking block {:?}", block_id);
+        self.service
+            .check_blocks(vec![block_id])
+            .expect("Failed to check block");
```

Comment:
In a future commit, we should fix these expects - there are many reasons this could fail, and probably shouldn't be considered a `panic` situation.

Reply:
Agreed. I think more experience with writing engines will help us to figure out what to do with various error situations. One solution might be to limit the possible errors that an engine can be confronted with by having the service do more error handling, e.g. resending the request if the response status is `NOT_READY`.